File: stone_lib/src/main/java/com/github/klee0kai/stone/types/wrappers/LazyProvide.java
package com.github.klee0kai.stone.types.wrappers;

public class LazyProvide<T> implements Ref<T> {

    private T value = null;
    private Ref<T> call;

    public LazyProvide(Ref<T> call) {
        this.call = call;
    }

    @Override
    public T get() {
        if (value != null)
            return value;
        return value = call.get();
    }
}

File: stone_multiplatform/src/commonMain/kotlin/stone/wrappers/LazyProvide.kt
package stone.wrappers

/**
 * Providing an object with lazy providing.
 * The object will be provided the first time it is used.
 * <pre>`ㅤ@Component
 * interface Component {
 *
 * LazyProvide<WelcomePresenter> presenter();
 *
 * }
`</pre> *
 */
class LazyProvide<T>(private val call: stone.wrappers.Ref<T?>) : stone.wrappers.Ref<T?> {
    private var value: T? = null

    override fun get(): T? {
        if (value != null) return value
        return call.get().also { value = it }
    }
}
